stl/inc/chrono

```diff
@@ -2209,6 +2221,2949 @@ namespace chrono {
+            if (!_Consistent || !_Apply_duration_fields(_Dur, true)) {
+                return false;
+            }
+
+            const _LeapSecondRep _Original_leap{_Leap};
+            if (_Leap == _LeapSecondRep::_File_time) {
+                const int _Year{*_Century + *_Two_dig_year};
+                _Leap = _Year <= 2016 ? _LeapSecondRep::_Negative : _LeapSecondRep::_All;
+            }
+
+            if (_Second > (_Leap == _LeapSecondRep::_All ? 60 : 59)) {
+                return false;
+            }
+
+            // A distinction has to be made here between clocks that tick monotonically, even around a positive leap
+            // second (everything except system_clock) and ones that have a special representation for leap seconds
+            // (utc_clock and, for negative leap seconds, system_clock). gps_clock and tai_clock, in particular, always
+            // monotonically count 86,400 seconds/day. The correct tick count, therefore, can be determined without
+            // knowing whether any leap seconds have occurred, and there aren't any invalid times due to negative leap
+            // second deletions.
+            //
+            // system_clock also has 86,400 seconds/day, but observes negative leap seconds by skipping them in its tick
+            // count. So leap second data is needed to check for a valid time, but not to calculate the tick count.
+
+            if (_Leap != _LeapSecondRep::_None) {
+                if (_Hour_24 == 23 && _Minute == 59 && _Second >= 59) {
+                    // It's possible that the parsed time doesn't exist because (a) _Seconds == 60 and there *isn't* a
+                    // leap second insertion or (b) _Seconds == 59 and there *is* a leap second subtraction.
+
+                    // Check for quick exit.
+                    const auto& _Tzdb{_CHRONO get_tzdb()};
+                    if (_Leap == _LeapSecondRep::_Negative && _Tzdb._All_ls_positive) {
+                        return true;
+                    }
+
+                    const bool _Possible_insertion{_Second == 60};
+                    const sys_seconds _Target_sys_time{
+                        _CHRONO floor<seconds>(_Dur) + (_Possible_insertion ? seconds{0} : seconds{1})};
+                    const auto& _Ls_vector{_Tzdb.leap_seconds};
+                    const auto _It{_STD lower_bound(_Ls_vector.begin(), _Ls_vector.end(), _Target_sys_time)};
+                    const bool _Match_leap{_It != _Ls_vector.end() && *_It == _Target_sys_time};
+
+                    // Condition for a good parse: (_Seconds == 59 && !_Match_leap) || (_Match_leap &&
+                    // _It->_Is_positive()). Below is the inverse of this.
+                    if (!(_Match_leap ? _It->_Positive() : !_Possible_insertion)) {
+                        return false;
+                    }
+                }
+
+                if (_Leap == _LeapSecondRep::_All) {

```

Comment:
Should that be an `else if` or can both of those branches be taken?

Reply:
Perhaps @MattStephanson can comment. That would be a major logic change, so I wouldn't want to touch this code without a crystal-clear understanding of what's happening (the only changes I am comfortable making without fully understanding code are those that are provably correct in isolation, like applying De Morgan's Laws). As far as I can tell, `else if` would be incorrect here - the first branch `if (_Hour_24 == 23 && _Minute == 59 && _Second >= 59) {` is testing the parsed time, and this branch `if (_Leap == _LeapSecondRep::_All) {` is testing the behavior of the clock (`_All` indicates `utc_clock` as commented above). I believe it is easily possible for the 23:59:59/60 condition to happen for `utc_clock`.

(As an aside, this is a good example of why we follow the convention "newline between non-chained `if`-statements", so it's clear when branches are independent.)

Reply:
@StephanTLavavej is right. The first branch is strictly for error-checking the seconds field (60 is valid only during a positive leap second insertion, 59 is not valid during a negative insertion). It's only necessary if the parsed time is 23:59:(59|60).

The second `if` is for offsetting the tick count based on leap seconds. It only applies to clocks that don't count exactly 86,400 seconds/day, i.e. only `utc_clock`. Also, it has to be done for every parsed `time_point` because the offset persists for every time after the first leap second insertion. (It might be possible to optimize this by skipping for times before the first leap second in 1972, but I haven't thought about it before now.)

To be concrete, this test is *supposed* to fail if you fuse the branches with an `else`:
```cpp
utc_seconds ut_ref = utc_clock::from_sys(sys_days{1d / July / 1972y}) - 1s; // leap second insertion
test_parse("june 30 23:59:60 1972", "%c", ut);
assert(ut == ut_ref);
```
The first branch is taken to ensure that the parsed time really is a leap second insertion, but if you skip the second branch you won't get any leap second offset. This particular test would work, though, because the +1s adjustment in `utc_clock::from_sys` happens to cancel `_Dur -= _CHRONO duration_cast<_DurationType>(seconds{1});`, which is equivalent to skipping the second branch entirely. It would be better to test one of the later leap seconds, where the accumulated offset is greater than 1s.

Reply:
I pushed an additional test (selecting 1992 since it was also a June 30 leap second year), with your rationale as a comment. Thanks!